pkg/scheduler/scheduler.go

```diff
@@ -347,6 +347,16 @@ func updatePod(client clientset.Interface, pod *v1.Pod, condition *v1.PodConditi
 	if nominatedNode != "" {
 		podCopy.Status.NominatedNodeName = nominatedNode
 	}
+	// See issue: #94626. This is to handle the case with empty name field in spec.ImagePullSecrets
+	// which is also the patchMergeKey and would fail the two-way merge patch generation and prevent scheduler from updating
+	// the pod conditions
+	var pullSecrets []v1.LocalObjectReference
```

Comment:
This is solving the symptom, rather than the root cause.

The root cause is that the produced 2-way merge patch includes changes to the ImagePullSecrets where it shouldn't.

Why? We are serializing the same values for secrets. Can you show the output of the marshaling and the produced patch?

Reply:
The root-cause is that producing the two merge way patch fails due to the missing`name` field. I'd expect modifying the strategic merge patch code to clear empty slices would be a breaking change - hence this proposal.

https://github.com/kubernetes/kubernetes/blob/06dfe683cefe06ee1cd09ab14b9c890b1c651786/staging/src/k8s.io/api/core/v1/types.go#L3072-L3074

This pod spec is valid on create:
```
...
imagePullSecrets:
- {}
...
```
Inside `PatchPod`:

oldData:
```
{"metadata":{"name":"pod1","namespace":"ns","creationTimestamp":null},"spec":{"containers":null,"imagePullSecrets":[{}]},"status":{}}
```
newData:
```
{"metadata":{"name":"pod1","namespace":"ns","creationTimestamp":null},"spec":{"containers":null,"imagePullSecrets":[{}]},"status":{"conditions":[{"type":"PodScheduled","status":"False","lastProbeTime":null,"lastTransitionTime":"2021-06-24T15:56:56Z"}]}}
```

Calling CreateTowoWayMergePatch(oldData, newData) will fail with:

```
map: map[] does not contain declared merge key: name
```

One way to avoid the patch creation failure is to drop the imagePullSecrets from the new pod which is what this change is doing:

With the change, newData becomes:
```
{"metadata":{"name":"pod1","namespace":"ns","creationTimestamp":null},"spec":{"containers":null},"status":{"conditions":[{"type":"PodScheduled","status":"False","lastProbeTime":null,"lastTransitionTime":"2021-06-24T16:03:26Z"}]}}
```

This wouldn't be an issue if that "bad input" was blocked on creates but that ship has long sailed and it seems the way it will be addressed is through warnings (https://github.com/kubernetes/kubernetes/issues/94626) to be backwards compatible.

I'm open to other suggestions for fixing that here.

Reply:
I always realized that `ClearNominatedNodeName` will call PatchPod and will likely hit the same issue so I can move that fix to happen inside `PatchPod` (which seems slightly odd to have in there), once we settle on an approach.

Reply:
Thanks for the context (#94626)

`PatchPod` is within the scheduler package, so it might be fine to do changes there. The scheduler is only concerned very few things, in terms of patching: conditions and nominated nodename, both part of the status.

I think a better alternative is to do some kind of allow list: instead of doing a DeepCopy of the entire pod, we just do it for the status. And we use the statuses for the creating the patch. Then, you probably have to add some extra formatting to create the Pod patch.

Reply:
Good point. That would be a lot cleaner since we only care about status. I'll adjust the change.